File: rd-kt/rd-framework/src/jvmTest/kotlin/com/jetbrains/rd/framework/test/cases/contexts/ContextWithExtTest.kt
package com.jetbrains.rd.framework.test.cases.contexts

import com.jetbrains.rd.framework.RdContext
import com.jetbrains.rd.framework.test.util.RdFrameworkTestBase
import demo.DemoModel
import demo.extModel
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class ContextWithExtTest : RdFrameworkTestBase() {

    @ParameterizedTest
    @ValueSource(booleans = [true, false])
    fun testExtPreserveContextOnLateConnect(useHeavyContext: Boolean) {
        println("useHeavyContext: $useHeavyContext")
        val context = if(useHeavyContext) ContextsTest.TestKeyHeavy else ContextsTest.TestKeyLight
        clientProtocol.serializers.register(RdContext.marshallerFor(context))
        serverProtocol.contexts.registerContext(context)
        
        val serverModel = DemoModel.create(serverLifetime, serverProtocol)
        val clientModel = DemoModel.create(clientLifetime, clientProtocol)

        val serverExt = serverModel.extModel

        val fireValues = listOf("a", "b", "c")

        setWireAutoFlush(false)

        fireValues.forEach {
            context.value = it
            serverExt.checker.fire(Unit)
            context.value = null
        }

        var numReceives = 0
        val receivedContexts = mutableSetOf<String>()

        val clientExt = clientModel.extModel
        clientExt.checker.advise(clientLifetime) {
            numReceives++
            receivedContexts.add(context.value ?: "null")
        }

        setWireAutoFlush(true)

        assertEquals(3, numReceives)
        assertEquals(fireValues.toSet(), receivedContexts)
    }
}

File: rd-net/Test.RdFramework/Contexts/ContextWithExtTest.cs
using System.Collections.Generic;
using JetBrains.Collections.Viewable;
using JetBrains.Rd;
using NUnit.Framework;
using Test.RdFramework.Interning;

namespace Test.RdFramework.Contexts
{
  public class ContextWithExtTest : RdFrameworkTestBase
  {
    [Datapoint]
    public static bool TrueValue = true;

    [Datapoint]
    public static bool FalseValue = false;
    
    [Theory]
    public void TestExtPreserveContextOnLateConnect(bool useHeavyContext)
    {
      var context = useHeavyContext
        ? RdContextBasicTest.TestKeyHeavy.Instance
        : (RdContext<string>) RdContextBasicTest.TestKeyLight.Instance;
      
      context.RegisterOn(ClientProtocol.Serializers);
      context.RegisterOn(ServerProtocol.Contexts);

      var serverModel = new InterningRoot1(LifetimeDefinition.Lifetime, ServerProtocol);
      var clientModel = new InterningRoot1(LifetimeDefinition.Lifetime, ClientProtocol);

      var serverExt = serverModel.GetOrCreateExtension("test", () => new InterningExt());

      var fireValues = new[] { "a", "b", "c" };

      ServerWire.AutoTransmitMode = false;
      ClientWire.AutoTransmitMode = false;
      
      foreach (var fireValue in fireValues)
      {
        context.Value = fireValue;
        serverExt.Root.Value = new InterningExtRootModel();
        context.Value = null;
      }

      var numReceives = 0;
      var receivedContexts = new HashSet<string>();

      var clientExt = clientModel.GetOrCreateExtension("test", () => new InterningExt());
      clientExt.Root.AdviseNotNull(LifetimeDefinition.Lifetime, _ =>
      {
        numReceives++;
        receivedContexts.Add(context.Value);
      });

      while (ClientWire.HasMessages || ServerWire.HasMessages)
      {
        ClientWire.TransmitAllMessages();
        ServerWire.TransmitAllMessages();
      }

      Assert.AreEqual(3, numReceives);
      Assert.AreEqual(fireValues, receivedContexts);
    }
  }
}